### PR TITLE
Add fields to configure notification sending in the deployment config

### PR DESCRIPTION
**Pull request by ono-max**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes part of #2342.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/config/deployment.go [hunk 1]
```diff
@@ -370,6 +372,21 @@ type SecretEncryption struct {
...
 
+// Notification represents the way to send to users.
+type Notification struct {
+	// List of users to be notified for each event.
+	Mentions []NotificationMention `json:"mentions"`
+}
+
+type NotificationMention struct {
+	// The event to be notified to users
+	Event model.Stage `json:"event"`
```
Comment (nakabonne) [inline pkg/config/deployment.go:383]: Possibly we can use `model.NotificationEventType`

Comment (nakabonne) [reply pkg/config/deployment.go:383]: But actually `model.NotificationEventType` is int32. So you can simple use `string` here. Then let's add a `Validate()` function to check if it's within `model.NotificationEventType_value` or not.

File: pkg/config/deployment.go [hunk 2]
```diff
@@ -44,6 +44,8 @@ type GenericDeploymentSpec struct {
 	Timeout Duration `json:"timeout,omitempty" default:"6h"`
 	// List of encrypted secrets and targets that should be decoded before using.
 	Encryption *SecretEncryption `json:"encryption"`
+	// Notification to be sent to users via Slack or email
```
Comment (nghialv) [inline pkg/config/deployment.go:47]: nit: `Additional configuration used while sending notification to external services.`

File: pkg/config/deployment.go [hunk 3]
```diff
@@ -370,6 +372,21 @@ type SecretEncryption struct {
...
 }
 
+// Notification represents the way to send to users.
+type Notification struct {
+	// List of users to be notified for each event.
+	Mentions []NotificationMention `json:"mentions"`
+}
+
+type NotificationMention struct {
+	// The event to be notified to users
```
Comment (nghialv) [inline pkg/config/deployment.go:382]: nit: Add the `.` at end of the line.

File: pkg/config/deployment.go (hunk 1 above)
Comment (ono-max) [reply pkg/config/deployment.go]: 7e2498b

File: pkg/config/deployment.go [hunk 4]
```diff
@@ -370,6 +372,21 @@ type SecretEncryption struct {
 	DecryptionTargets []string `json:"decryptionTargets"`
 }
 
+// Notification represents the way to send to users.
+type Notification struct {
```
[the remark this answers is not in the document]
Comment (ono-max) [reply pkg/config/deployment.go]: fb289b7

File: pkg/config/deployment.go (hunk 2 above)
Comment (ono-max) [reply pkg/config/deployment.go]: 26713d1

File: pkg/config/deployment.go (hunk 3 above)
Comment (ono-max) [reply pkg/config/deployment.go]: 26713d1

**Comment by nakabonne** [comment]
@ono-max Could you please stop force pushing as much as possible so that we can follow what was changed exactly? (sorry we haven't said that yet)

**Comment by nakabonne** [comment]
I'm not sure why it shows you that but you can ignore https://github.com/pipe-cd/pipe/pull/2503#issuecomment-924530315

File: pkg/config/deployment.go [hunk 5]
```diff
@@ -381,3 +383,27 @@ func (e *SecretEncryption) Validate() error {
...
+	Email []string `json:"email"`
+}
+
+func (n *NotificationMention) Validate() error {
+	for k := range model.NotificationEventType_value {
+		if n.Event == k {
+			return nil
+		}
+	}
+	return fmt.Errorf("the field name %s must be NotificationEventType", n.Event)
```
Comment (nakabonne) [inline pkg/config/deployment.go:408]: Nice to have the field name like:
```suggestion
	return fmt.Errorf("event %q is incorrect as NotificationEventType", n.Event)
```

**Comment by nghialv** [comment]
nit: FYI, the pull request title is used to generate our release changelog so we need a message that is easier to understand from the user's point of view.

> Define Notification struct to unmarshal the notification key of YAML to Config struct.

So how about this and don't use the trailing dot.

> Add fields to configure notification sending in the deployment config

**Comment by ono-max** [comment]
@nghialv 

Thanks, I've updated it.

**Comment by nakabonne** [comment]
There you go!
/approve

**Pull request merged**